test/common/upstream/hds_test.cc

```diff
@@ -0,0 +1,179 @@
+#include "envoy/service/discovery/v2/hds.pb.h"
+
+#include "common/ssl/context_manager_impl.h"
+#include "common/stats/stats_impl.h"
+#include "common/upstream/health_discovery_service.h"
+
+#include "test/common/upstream/hds_helper.h"
+#include "test/mocks/event/mocks.h"
+#include "test/mocks/grpc/mocks.h"
+#include "test/mocks/upstream/mocks.h"
+#include "test/test_common/utility.h"
+
+#include "gmock/gmock.h"
+#include "gtest/gtest.h"
+
+using testing::InSequence;
+using testing::Invoke;
+using testing::NiceMock;
+using testing::Return;
+using testing::_;
+
+namespace Envoy {
+namespace Upstream {
+
+class HdsTest : public testing::Test {
+public:
+  HdsTest()
+      : retry_timer_(new Event::MockTimer()), server_response_timer_(new Event::MockTimer()),
+        async_client_(new Grpc::MockAsyncClient()) {
+    node_.set_id("foo");
+  }
+
+  void createHdsDelegate() {
+    InSequence s;
+    EXPECT_CALL(dispatcher_, createTimer_(_)).WillOnce(Invoke([this](Event::TimerCb timer_cb) {
+      retry_timer_cb_ = timer_cb;
+      return retry_timer_;
+    }));
+    EXPECT_CALL(dispatcher_, createTimer_(_)).WillOnce(Invoke([this](Event::TimerCb timer_cb) {
+      server_response_timer_cb_ = timer_cb;
+      return server_response_timer_;
+    }));
+    hds_delegate_.reset(new HdsDelegate(node_, stats_store_, Grpc::AsyncClientPtr(async_client_),
+                                        dispatcher_, runtime_, stats_store_, ssl_context_manager_,
+                                        secret_manager_, random_, test_factory_));
+  }
+
+  envoy::api::v2::core::Node node_;
+  Event::MockDispatcher dispatcher_;
+
+  Stats::IsolatedStoreImpl stats_store_;
+
+  HdsDelegatePtr hds_delegate_;
+  TestHdsInfoFactory test_factory_;
+
+  Event::MockTimer* retry_timer_;
+  Event::TimerCb retry_timer_cb_;
+  Event::MockTimer* server_response_timer_;
+  Event::TimerCb server_response_timer_cb_;
+
+  std::unique_ptr<envoy::service::discovery::v2::HealthCheckSpecifier> message;
+  Grpc::MockAsyncStream async_stream_;
+  Grpc::MockAsyncClient* async_client_;
+  Runtime::MockLoader runtime_;
+  Ssl::ContextManagerImpl ssl_context_manager_{runtime_};
+  Secret::MockSecretManager secret_manager_;
+  NiceMock<Runtime::MockRandomGenerator> random_;
+};
+
+TEST_F(HdsTest, TestProcessMessageEndpoints) {
```

Comment:
This is great. I notice above you have added some captures for the timer CBs; do you want some unit tests that make use of these? You can see how to use them in the subscription unit tests (I can point you to them tomorrow, I'm on a bad plane wifi right now).